L70-node-proto-loader-type-generator.md

```diff
@@ -0,0 +1,216 @@
+}
+
+export interface ProtoGrpcType {
+  package_name: {
+    subpackage_name: {
+      EnumName: EnumTypeDefinition
+      MessageName: MessageTypeDefinition
+      ServiceName: SubtypeConstructor<typeof grpc.Client, _package_name_subpackage_name_ServiceNameClient> & { service: ServiceDefinition }
+    }
+  }
+}
+```
+
+## Rationale
+
+The goal is to have type information available when editing and building code that uses `@grpc/proto-loader` to load `.proto` files at runtime. It is not possible to infer this type information from the interfaces currently provided by this library because TypeScript cannot infer type information from `.proto` files. The only other option is to generate this type information separately. The recommended usage of the `@grpc/proto-loader` library is to call `load` or `loadSync` and then pass the result of that to `grpc.loadPackageDefinition`, so the type of the final result of that will be useful to most users. The output of `load` and `loadSync` are objects with runtime data that describes the resulting types, so the types of those objects are significantly less useful, and the final type that we will generate here cannot be effectively inferred from those types.
+
+The type parameters `grpc.Client`, `grpc.Metadata`, and `grpc.CallOptions` are all needed because each of those types is used in the resulting type, and none of them can be inferred from the others. The type `grpc.Metadata` also impacts the resulting type, but its type is the same across the two implementations so it can be defined independently of them.
+
+These alternatives were considered to generating files in a directory structure based on the protobuf package structure:
+
+ - **Generate files in a directory structure mirroring the directory structure of the input files, all relative to the output directory.** This introduces the complexity of handling unusual import paths, including absolute paths and paths with `..`.
+ - **Generate files corresponding to the input files, all directly in the output directory.** This greatly increases the risk of filename conflicts.
+ - **Generate a single file with all of the types.** This creates a large, unwieldy file, making it hard for a human to evaluate changes. This can also greatly increase code duplication if multiple files need to be generated.
+
+The gRPC implementation is specified at build time because the types depend on types from that library and the user should know at that point which implementation they are using. An alternative is to use generics to insert the types in a different way, but that increases the complexity of both the generated code and the code that uses it for relatively little gain.
+
+Two types are generated for each message definition to represent that at runtime the library is permissive with what it accepts, but stricter with what it outputs. Having those stricter output types allows the user to get tigher type guarantees where possible.

```

Comment:
This sentence is  hard to read. Do you mean the following: 

"Two types are generated for each message definition. To represent the message at runtime, the library (tool?) is .... "

Reply:
What I'm trying to say is that the library is permissive about what types it accepts as input, but we can be more specific about what types it outputs and the two different types reflect that.